### DISCUSSION TITLE
Data Grid rounded corners?

### DISCUSSION THREAD

**Discussion by BrunoDenisov** [AvaloniaUI/Avalonia#17766] [General]
I'm wondering if anyone has done so that only the outer corners of the data grid are rounder when using corner radius

**Comment by LaurentInSeattle** [comment]
You can easily place any visual in a border, give it rounded corners and have it clip its content.

Another way, Add CornerRadius to Image (for example): 

`
public class RoundedImage : Image
{
    public static readonly AttachedProperty<double> CornerRadiusProperty =
        AvaloniaProperty.RegisterAttached<RoundedImage, double>(
            "CornerRadius", typeof(RoundedImage), 8);

    public static void SetCornerRadius(AvaloniaObject element, double parameter)
        => element.SetValue(CornerRadiusProperty, parameter);

    public static double GetCornerRadius(AvaloniaObject element)
        => element.GetValue(CornerRadiusProperty);

    protected override Size MeasureOverride(Size availableSize)
    {
        IImage? source = this.Source;
        if (source is not null)
        {
            Size result = 
                this.Stretch.CalculateSize(availableSize, source.Size, this.StretchDirection);
            this.Clip =
                new RectangleGeometry(
                    new Rect(0, 0, result.Width, result.Height),
                    RoundedImage.GetCornerRadius(this),
                    RoundedImage.GetCornerRadius(this));
            return result;
        }

        return base.MeasureOverride(availableSize);
    }
}
`

**Comment by msneijders** [comment]
You mean like this?

![image](https://github.com/user-attachments/assets/b3d56000-f606-49f3-9127-8bc50417a47c)

Which 'just works' was it not that `ClipToBounds` is not set to the border in the built-in DataGrid ControlTheme, making it look like so:

![image](https://github.com/user-attachments/assets/a835d076-37db-4ed9-89f2-a3833b33f6e2)

Which I fixed by doing this:

```xaml
<ControlTheme
    x:Key="MyDataGrid"
    BasedOn="{StaticResource {x:Type DataGrid}}"
    TargetType="DataGrid">
    <Style Selector="^ /template/ Border#DataGridBorder">
        <Setter Property="ClipToBounds" Value="True" />
    </Style>
</ControlTheme>

<DataGrid
   CornerRadius="50"
   Theme={StaticResource MyDataGrid}
   ClipToBounds="True" <!-- but this one is not bound to the border, so not working by default -->
/>
```

It might be an oversight in the default DataGrid ControlTheme -- i.e. the default ControlTheme for ListBox does bind the ClipToBounds to its Border.

Other easier workaround, just surround the DataGrid with another `<Border CornerRadius="50" ClipToBounds="True">`.
